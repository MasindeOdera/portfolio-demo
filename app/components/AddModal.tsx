import { useState } from 'react';
import styled from 'styled-components';
import ModalContainer from './ModalContainer';
import { theme } from '../styles/theme';

type NewProject = {
  project: string;
  image: string;
  description: string;
};

interface AddModalProps {
  onCancel: () => void;
  onAdd: (newProject: NewProject) => void;
}

const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: 10px;
`;

const Label = styled.label`
  font-size: 0.9rem;
  font-weight: bold;
  color: ${theme.colors.primary};
  margin-bottom: 5px;
`;

const InfoText = styled.p`
  font-size: 0.9rem;
  color: ${theme.colors.primary};
  font-style: italic;
  margin-bottom: 10px;
`;

const Input = styled.input`
  padding: 10px;
  border-radius: 5px;
  border: 1px solid ${theme.colors.detail};
  font-size: 1rem;
`;

const TextArea = styled.textarea`
  padding: 10px;
  border-radius: 5px;
  border: 1px solid ${theme.colors.detail};
  font-size: 1rem;
  height: 80px;
  resize: none;
`;

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin-top: 20px;
`;

const Button = styled.button`
  width: 45%;
  padding: 10px 20px;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  font-weight: bold;

  &:nth-child(1) {
    background-color: ${theme.colors.success};
    color: ${theme.colors.text};
  }

  &:nth-child(2) {
    background-color: ${theme.colors.background};
    color: ${theme.colors.text};
  }

  &:hover {
    opacity: 0.9;
  }
`;

export default function AddModal({ onCancel, onAdd }: AddModalProps) {
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');

  const generateRandomImage = () => {
    const categories = ['nature', 'business', 'technology', 'fashion', 'transport', 'food', 'animals'];
    const randomCategory = categories[Math.floor(Math.random() * categories.length)];
    const randomImageUrl = `https://loremflickr.com/640/480/${randomCategory}?random=${Math.random()}`;
    return randomImageUrl;
  };

  const handleAdd = () => {
    if (name && description) {
      const randomImage = generateRandomImage();
      onAdd({ project: name, description, image: randomImage });
    }
  };

  return (
    <ModalContainer title="Add New Project">
      <FormContainer>
        <Label htmlFor="title">Title</Label>
        <Input 
          id="title" 
          value={name} 
          onChange={(e) => setName(e.target.value)} 
        />
        
        <Label htmlFor="description">Description</Label>
        <TextArea 
          id="description" 
          value={description} 
          onChange={(e) => setDescription(e.target.value)} 
        />

        <InfoText>
          Your project screenshot will be automatically generated from a random image.
        </InfoText>
      </FormContainer>

      <ButtonContainer>
        <Button onClick={handleAdd}>Add</Button>
        <Button onClick={onCancel}>Cancel</Button>
      </ButtonContainer>
    </ModalContainer>
  );
}
