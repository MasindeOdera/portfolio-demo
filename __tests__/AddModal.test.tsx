import { render, screen, fireEvent } from '@testing-library/react';
import AddModal from '@/app/components/AddModal';

describe('AddModal Component', () => {
  const mockOnCancel = jest.fn();
  const mockOnAdd = jest.fn();

  beforeEach(() => {
    render(<AddModal onCancel={mockOnCancel} onAdd={mockOnAdd} />);
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<AddModal onCancel={mockOnCancel} onAdd={mockOnAdd} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders form inputs correctly', () => {
    expect(screen.getByLabelText(/Title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/Description/i)).toBeInTheDocument();
  });

  test('updates input values correctly', () => {
    const titleInput = screen.getByLabelText(/Title/i);
    const descriptionInput = screen.getByLabelText(/Description/i);

    fireEvent.change(titleInput, { target: { value: 'New Project' } });
    fireEvent.change(descriptionInput, { target: { value: 'This is a new project description.' } });

    expect(titleInput).toHaveValue('New Project');
    expect(descriptionInput).toHaveValue('This is a new project description.');
  });

  test('calls onCancel when Cancel button is clicked', () => {
    const cancelButton = screen.getByText(/Cancel/i);
    fireEvent.click(cancelButton);
    expect(mockOnCancel).toHaveBeenCalled();
  });
});
