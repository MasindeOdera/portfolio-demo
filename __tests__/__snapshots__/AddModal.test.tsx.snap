// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddModal Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-blHHSb bkqoCW"
  >
    <h2
      class="sc-gtLWhw jidGLU"
    >
      Add New Project
    </h2>
    <div
      class="sc-egkSDF leRVCI"
    >
      <div
        class="sc-fAUdSK icfotn"
      >
        <label
          class="sc-dntaoT gRrdXS"
          for="title"
        >
          Title
        </label>
        <input
          class="sc-ghWlax fxkiwY"
          id="title"
          value=""
        />
        <label
          class="sc-dntaoT gRrdXS"
          for="description"
        >
          Description
        </label>
        <textarea
          class="sc-kLhKbu hRfzoR"
          id="description"
        />
        <p
          class="sc-ivxoEo cqzXEI"
        >
          Your project screenshot will be automatically generated from a random image.
        </p>
      </div>
      <div
        class="sc-ixGGxD cBuzsV"
      >
        <button
          class="sc-khLCKb ewVQdg"
        >
          Add
        </button>
        <button
          class="sc-khLCKb ewVQdg"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
